doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+
+$$
+i_1 : t_1 \ , \ldots \ , i_n : i_n | v
+$$
+
+with each $t_k$-typed field indexed by label $i_k$ for $1 \le k \le n$ and the
+free row variable $v$, then
+
+$$
+\\{i_1 : t_1 \ , \ldots \ , i_n : i_n | v\\}
+$$
+
+is an open record conjoining the fields, and 
+
+$$
+\langle i_1 : t_1 \ , \ldots \ , i_n : i_n | v \rangle
+$$
+
+is an open sum type presenting the fields as (mutually exclusive) alternatives.
+Both types are extensible by substituting $v$ with another (possibly open row).
+To represent a closed row, we omit the trailing $| v$.
+
+### Quint's current type system
+
+The [current type system supported by quint][type-system] is based on a
+simplified version of the constraint-based system presented in ["Complete and
+Decidable Type Inference for GADTs"][gadts] augmented with extensible
+(currently, just) records based on "Extensible Records with Scoped Labels". A
+wrinkle in this genealogy is that quint's type system includes neither GADTs nor
+scoped labels (and even the extensiblity supported for records is limited).
+Moreover, due to their respective foci, neither of the referenced papers
+includes a formalization the complete statics for product types or sum types,
+and while we have implemented support for product types in quint, we don't have
+our typing rules recorded.
+
+## Statics
+
+This section discusses the typing judgements that will allow us to statically
+verify correct introduction and elimination of expressions that are variants of
+a sum type. The following considerations have informed the structure in which
+the proposed statics are discussed:
+
+- Since sum-types are dual to product types, I consider their complementary
+  typing rules together: first I will present the relevant rule for product
+  types, then propose the complementary rule for sum types. This should help
+  maintain consistency between the two kinds of typing judgements and ensure
+  our implementations of both harmonize.
+- Since we don't have our existing product formation or elimination rules
+  described separate from the implementation, transcribing them here can serve
+  to juice our intuition, supplement our design documentation, and perhaps give
+  opportunity for refinement.
+- Since our homegrown type system has some idiosyncrasies that can obscure the
+  essence of the constructs under discussion, I precede the exposition of each
+  rule with a text-book example adapted from [Practical Foundations for
+  Programming Languages][PFPL]. This is only meant as a clarifying touchstone.
+
+### Eliminating products and introducing sums
+
+The elimination of products via projection and the introduction of sums via
+injection are the simplest of the two pairs of rules.
+
+#### Projection
+
+Here is a concrete example of projecting a value out of a record using our
+current syntax:
+
+```quint
+val r : {a:int} = {a:1}
+val ex : int = r.a
+// Or, using our exotic field operator, which is currently the normal form
+val ex_ : int = r.field("a")
```

Comment:
By reading this, I realized that our normal form is just the abstract syntax tree (AST) that is exposed in the syntax. For many of the Quint operators, the normal form (AST) coincides with the surface syntax. For some cases, like `r.field("a")`, the surface syntax is different. I guess, the same would happen if we tried to express Quint syntax in Lisp, though we could use atoms there, e.g., `(field r 'a)`.

Reply:
It's not the AST really, because the AST doesn't have quint strings for labels. These operators are quite weird because the type of the return expression depends on the quint *value* supplied as an argument. I'm not sure we want that kind of dependently typed operator in the long term, but this is a broader question that is properly out of scope for this PR. :)